File: app/src/main/java/com/example/deltatask3/fragments/PokemonsFragment.java
package com.example.deltatask3.fragments;

import android.app.ActivityOptions;
import android.content.Intent;
import android.graphics.Canvas;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.SearchView;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.deltatask3.R;
import com.example.deltatask3.activities.PokemonDetailsActivity;
import com.example.deltatask3.adapters.PokemonAdapter;
import com.example.deltatask3.api.PokemonApi;
import com.example.deltatask3.database.Favourite;
import com.example.deltatask3.databinding.FragmentPokemonsBinding;
import com.example.deltatask3.utils.Pokemon;
import com.example.deltatask3.utils.SearchResult;
import com.example.deltatask3.viewmodels.AppViewModel;
import com.example.deltatask3.viewmodels.FavouriteViewModel;
import com.google.gson.Gson;
import com.muddzdev.styleabletoast.StyleableToast;

import java.util.ArrayList;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import it.xabaras.android.recyclerview.swipedecorator.RecyclerViewSwipeDecorator;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@AndroidEntryPoint
public class PokemonsFragment extends Fragment {

    private static final String TAG = "PokemonsFragment";
    private FragmentPokemonsBinding binding;

    private FavouriteViewModel favouriteViewModel;

    @Inject
    PokemonApi pokemonApi;

    private ArrayList<String> names;
    private ArrayList<Pokemon> allPokemons;
    private ArrayList<Favourite> favourites;
    private ArrayList<Pokemon> searchedPokemon;

    private PokemonAdapter pokemonAdapter;
    private LinearLayoutManager layoutManager;
    private int offset = 0;

    private SearchView searchView;

    private boolean loading = true, searching = false;

    public PokemonsFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentPokemonsBinding.inflate(inflater, container, false);
        setHasOptionsMenu(true);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppViewModel appViewModel = new ViewModelProvider(requireActivity()).get(AppViewModel.class);
        appViewModel.setCurrentTitle("Pokémon");
        favouriteViewModel = new ViewModelProvider(requireActivity(), ViewModelProvider.AndroidViewModelFactory.getInstance(requireActivity().getApplication())).get(FavouriteViewModel.class);

        names = new ArrayList<>();

        allPokemons = new ArrayList<>();
        favourites = new ArrayList<>();
        searchedPokemon = new ArrayList<>();

        buildRecyclerView();
        getMorePokemon();
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        favouriteViewModel.getAllFavourites().observe(getViewLifecycleOwner(), newFavourites -> {
            favourites.clear();
            favourites.addAll(newFavourites);
        });
    }

    private void buildRecyclerView() {
        binding.allPokemonList.setHasFixedSize(true);
        layoutManager = new LinearLayoutManager(requireContext());
        pokemonAdapter = new PokemonAdapter();
        pokemonAdapter.setPokemons(allPokemons);
        pokemonAdapter.setListener(this::showDetails);
        binding.allPokemonList.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0) {
                    int visibleItemCount = layoutManager.getChildCount();
                    int totalItemCount = layoutManager.getItemCount();
                    int pastVisibleItems = layoutManager.findFirstVisibleItemPosition();

                    if (loading) {
                        if ((visibleItemCount + pastVisibleItems) >= totalItemCount) {
                            loading = false;
                            Log.i(TAG, "onScrolled: LastItem");
                            paginate();
                        }
                    }
                }
            }
        });
        binding.allPokemonList.setLayoutManager(layoutManager);
        binding.allPokemonList.setAdapter(pokemonAdapter);

        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                Pokemon swipedPokemon = pokemonAdapter.getPokemonAt(viewHolder.getAdapterPosition());
                if (pokemonIsInFavourites(swipedPokemon)) {
                    pokemonAdapter.notifyItemChanged(viewHolder.getAdapterPosition());
                    StyleableToast.makeText(requireContext(), firstLetterToUppercase(swipedPokemon.getName()) + " is already in favourites", Toast.LENGTH_SHORT, R.style.ToastTheme).show();
                } else {
                    addToFavourites(viewHolder.getAdapterPosition(), swipedPokemon);
                }
            }

            @Override
            public void onChildDraw(@NonNull Canvas c, @NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, float dX, float dY, int actionState, boolean isCurrentlyActive) {
                new RecyclerViewSwipeDecorator.Builder(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                        .addSwipeRightBackgroundColor(Color.parseColor("#EB3939"))
                        .addSwipeRightActionIcon(R.drawable.favourite_icon)
                        .create()
                        .decorate();
                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);
            }
        }).attachToRecyclerView(binding.allPokemonList);
    }

    private boolean pokemonIsInFavourites(Pokemon pokemon) {
        for (Favourite favourite : favourites) {
            if (favourite.getPokemon().getId() == pokemon.getId())
                return true;
        }
        return false;
    }

    private void addToFavourites(int position, Pokemon pokemon) {
        StyleableToast.makeText(requireContext(), firstLetterToUppercase(pokemon.getName()) + " added to favourites", Toast.LENGTH_SHORT, R.style.ToastTheme).show();
        favouriteViewModel.insert(new Favourite(pokemon));
        names.remove(pokemon.getName());
        allPokemons.remove(pokemon);
        if (searching)
            searchedPokemon.remove(pokemon);
        pokemonAdapter.notifyItemRemoved(position);
    }

    private void paginate() {
        offset += 20;
        getMorePokemon();
    }

    private void getMorePokemon() {
        Call<SearchResult> pokemonSearchResultCall = pokemonApi.getPokemonWithOffsetAndLimit(offset, 20);
        pokemonSearchResultCall.enqueue(new Callback<SearchResult>() {
            @Override
            public void onResponse(Call<SearchResult> call, Response<SearchResult> response) {
                if (!response.isSuccessful()) {
                    Log.i(TAG, "onResponse: " + response);
                    return;
                }


                for (SearchResult.Result result : response.body().getResults()) {
                    names.add(result.getName());
                    allPokemons.add(new Pokemon(result.getName()));
                }

                loading = true;
                loadDetailsIntoPokemons(names);
            }

            @Override
            public void onFailure(Call<SearchResult> call, Throwable t) {
                Log.i(TAG, "showAllPokemon failed");
            }
        });
    }

    private void loadDetailsIntoPokemons(ArrayList<String> names) {
        for (String s : names) {
            Call<Pokemon> call = pokemonApi.getPokemonFromName(s);
            call.enqueue(new Callback<Pokemon>() {
                @Override
                public void onResponse(Call<Pokemon> call, Response<Pokemon> response) {
                    if (!response.isSuccessful()) {
                        Log.i(TAG, "onResponse: " + response);
                        return;
                    }

                    if (names.indexOf(s) >= 0)
                        allPokemons.set(names.indexOf(s), response.body());
                    if (searching)
                        searchPokemonByName(searchView.getQuery().toString().trim().toLowerCase());
                    pokemonAdapter.notifyDataSetChanged();
                }

                @Override
                public void onFailure(Call<Pokemon> call, Throwable t) {
                    Log.i(TAG, "showAllPokemon failed");
                }
            });
        }
    }

    private void searchPokemonByName(String name) {
        searchedPokemon.clear();

        for (Pokemon pokemon : allPokemons) {
            if (pokemon.getName().trim().contains(name))
                searchedPokemon.add(pokemon);
        }

        pokemonAdapter.setPokemons(searchedPokemon);
        pokemonAdapter.notifyDataSetChanged();
    }

    private void showDetails(int position, ImageView pokemonIv, TextView nameIv) {
        Pokemon pokemon = pokemonAdapter.getPokemonAt(position);
        if (pokemon.getId() != 0 && pokemon.getSprites() != null) {
            Intent intent = new Intent(requireActivity(), PokemonDetailsActivity.class);
            Gson gson = new Gson();
            String pokemonJson = gson.toJson(pokemon);
            intent.putExtra("pokemonJson", pokemonJson);
            Pair<View, String> imagePair = new Pair<>(pokemonIv, "pokemonImg");
            Pair<View, String> namePair = new Pair<>(nameIv, "pokemonName");
            ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(requireActivity(), imagePair, namePair);
            startActivity(intent, options.toBundle());
        }
    }

    private String firstLetterToUppercase(String string) {
        return string.substring(0, 1).toUpperCase() + string.substring(1);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        menu.clear();
        inflater.inflate(R.menu.search_menu, menu);
        MenuItem item = menu.findItem(R.id.search);
        searchView = (SearchView) item.getActionView();
        searchView.setQueryHint("Search Pokémon");
        searchView.setImeOptions(EditorInfo.IME_ACTION_DONE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.i(TAG, "submit");
                searching = true;
                if (query.length() == 0) {
                    searchedPokemon.clear();
                    pokemonAdapter.setPokemons(allPokemons);
                    pokemonAdapter.notifyDataSetChanged();
                } else
                    searchPokemonByName(query.toLowerCase().trim());
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                searching = true;
                if (newText.length() == 0) {
                    searchedPokemon.clear();
                    pokemonAdapter.setPokemons(allPokemons);
                    pokemonAdapter.notifyDataSetChanged();
                } else
                    searchPokemonByName(newText.toLowerCase().trim());
                return true;
            }
        });
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        binding = null;
    }
}

File: app/src/main/java/com/example/deltatask3/fragments/PokemonsFragment.kt
package com.example.deltatask3.fragments

import android.app.ActivityOptions
import android.content.Intent
import android.graphics.Canvas
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.util.Pair
import android.view.*
import android.view.inputmethod.EditorInfo
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.deltatask3.R
import com.example.deltatask3.activities.PokemonDetailsActivity
import com.example.deltatask3.adapters.PokemonAdapter
import com.example.deltatask3.api.PokemonApi
import com.example.deltatask3.database.Favourite
import com.example.deltatask3.databinding.FragmentPokemonsBinding
import com.example.deltatask3.firstLetterToUppercase
import com.example.deltatask3.models.Pokemon
import com.example.deltatask3.showSnackbarInMain
import com.example.deltatask3.viewmodels.FavouriteViewModel
import com.google.android.material.snackbar.Snackbar
import com.google.gson.Gson
import com.muddzdev.styleabletoast.StyleableToast
import dagger.hilt.android.AndroidEntryPoint
import it.xabaras.android.recyclerview.swipedecorator.RecyclerViewSwipeDecorator
import kotlinx.coroutines.*
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class PokemonsFragment : Fragment() {

    companion object {
        private const val TAG = "PokemonsFragment"
    }

    private var _binding: FragmentPokemonsBinding? = null
    private val binding get() = _binding!!

    private lateinit var favouriteViewModel: FavouriteViewModel

    @Inject
    lateinit var pokemonApi: PokemonApi

    private val names = ArrayList<String>()
    private val allPokemons = ArrayList<Pokemon>()
    private val favourites = ArrayList<Favourite>()
    private val searchedPokemon = ArrayList<Pokemon>()
    private val pokemonAdapter = PokemonAdapter(allPokemons) { position: Int, pokemonIv: ImageView, nameIv: TextView -> showDetails(position, pokemonIv, nameIv) }
    private lateinit var layoutManager: LinearLayoutManager

    private var offset = 0
    private lateinit var searchView: SearchView
    private var loading = true
    private var searching = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        _binding = FragmentPokemonsBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        favouriteViewModel = ViewModelProvider(requireActivity(), ViewModelProvider.AndroidViewModelFactory.getInstance(requireActivity().application)).get(FavouriteViewModel::class.java)
        buildRecyclerView()
        getPokemon()
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        favouriteViewModel.allFavourites.observe(viewLifecycleOwner, { newFavourites: List<Favourite>? ->
            favourites.clear()
            favourites.addAll(newFavourites!!)
        })
    }

    private fun buildRecyclerView() {
        binding.allPokemonList.setHasFixedSize(true)
        layoutManager = LinearLayoutManager(requireContext())
        binding.allPokemonList.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0) {
                    val visibleItemCount = layoutManager.childCount
                    val totalItemCount = layoutManager.itemCount
                    val pastVisibleItems = layoutManager.findFirstVisibleItemPosition()
                    if (loading) {
                        if (visibleItemCount + pastVisibleItems >= totalItemCount) {
                            loading = false
                            Log.i(TAG, "onScrolled: LastItem")
                            paginate()
                        }
                    }
                }
            }
        })
        binding.allPokemonList.layoutManager = layoutManager
        binding.allPokemonList.adapter = pokemonAdapter
        ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val swipedPokemon = pokemonAdapter.getPokemonAt(viewHolder.adapterPosition)
                if (pokemonIsInFavourites(swipedPokemon)) {
                    pokemonAdapter.notifyItemChanged(viewHolder.adapterPosition)
                    StyleableToast.makeText(requireContext(), firstLetterToUppercase(swipedPokemon.name) + " is already in favourites", Toast.LENGTH_SHORT, R.style.ToastTheme).show()
                } else {
                    addToFavourites(viewHolder.adapterPosition, swipedPokemon)
                }
            }

            override fun onChildDraw(c: Canvas, recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
                RecyclerViewSwipeDecorator.Builder(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
                        .addSwipeRightBackgroundColor(Color.parseColor("#EB3939"))
                        .addSwipeRightActionIcon(R.drawable.favourite_icon)
                        .create()
                        .decorate()
                super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)
            }
        }).attachToRecyclerView(binding.allPokemonList)
    }

    private fun pokemonIsInFavourites(pokemon: Pokemon): Boolean {
        for (favourite in favourites) {
            if (favourite.pokemon.id == pokemon.id) return true
        }
        return false
    }

    private fun addToFavourites(position: Int, pokemon: Pokemon) {
        StyleableToast.makeText(requireContext(), firstLetterToUppercase(pokemon.name) + " added to favourites", Toast.LENGTH_SHORT, R.style.ToastTheme).show()
        favouriteViewModel.insert(Favourite(pokemon))
        names.remove(pokemon.name)
        allPokemons.remove(pokemon)
        if (searching) searchedPokemon.remove(pokemon)
        pokemonAdapter.notifyItemRemoved(position)
    }

    private fun paginate() {
        offset += 20
        getPokemon()
    }

    private fun getPokemon() {
        lifecycleScope.launch(Dispatchers.IO) {
            val response = pokemonApi.getPokemon(offset, 20)
            if (!response.isSuccessful)
                showSnackbarInMain(requireView(), "Failed to Fetch Pokemon", Snackbar.LENGTH_SHORT)
            else {
                val results = response.body()!!.results
                val deferredList = ArrayList<Deferred<Pokemon>>()
                for (result in results)
                    deferredList.add(async { pokemonApi.getPokemon(result.name).body()!! })
                allPokemons.addAll(deferredList.awaitAll())
                loading = true
                withContext(Dispatchers.Main) {
                    binding.allPokemonList.setHasFixedSize(false)
                    pokemonAdapter.notifyItemRangeInserted(allPokemons.size - results.size, results.size)
                    binding.allPokemonList.setHasFixedSize(true)
                }
            }
        }
    }

    private fun searchPokemonByName(name: String) {
        searchedPokemon.clear()
        for (pokemon in allPokemons)
            if (pokemon.name.trim().contains(name)) searchedPokemon.add(pokemon)
        pokemonAdapter.pokemons = searchedPokemon
        pokemonAdapter.notifyDataSetChanged()
    }

    private fun showDetails(position: Int, pokemonIv: ImageView, nameIv: TextView) {
        val pokemon = pokemonAdapter.getPokemonAt(position)
        if (pokemon.id != 0 && pokemon.sprites != null) {
            val intent = Intent(requireActivity(), PokemonDetailsActivity::class.java)
            val gson = Gson()
            val pokemonJson = gson.toJson(pokemon)
            intent.putExtra("pokemonJson", pokemonJson)
            val imagePair = Pair<View, String>(pokemonIv, "pokemonImg")
            val namePair = Pair<View, String>(nameIv, "pokemonName")
            val options = ActivityOptions.makeSceneTransitionAnimation(requireActivity(), imagePair, namePair)
            startActivity(intent, options.toBundle())
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        menu.clear()
        inflater.inflate(R.menu.search_menu, menu)
        val item = menu.findItem(R.id.search)
        searchView = item.actionView as SearchView
        searchView.queryHint = "Search Pokémon"
        searchView.imeOptions = EditorInfo.IME_ACTION_DONE
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                Log.i(TAG, "submit")
                searching = true
                if (query.isEmpty()) {
                    searchedPokemon.clear()
                    pokemonAdapter.pokemons = allPokemons
                    pokemonAdapter.notifyDataSetChanged()
                } else searchPokemonByName(query.toLowerCase(Locale.ROOT).trim())
                return false
            }

            override fun onQueryTextChange(newText: String): Boolean {
                searching = true
                if (newText.isEmpty()) {
                    searchedPokemon.clear()
                    pokemonAdapter.pokemons = allPokemons
                    pokemonAdapter.notifyDataSetChanged()
                } else searchPokemonByName(newText.toLowerCase(Locale.ROOT).trim())
                return true
            }
        })
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}